### PR TITLE
Fix Monero address validation (#1652)

### PR DESCRIPTION
**Pull request by Technohacker**
Add a RegexAddressValidator and tests

### CODE REVIEW

**Review by matt2134711 — CHANGES_REQUESTED** [review]
[Monero addresses are encoded in base58](https://github.com/monero-project/monero/blob/master/src/common/base58.cpp), so using `[0-9A-Za-z]` would still let some bad characters through (`0`, `I`, `O`, and `l` are never in an address).

Also the last few bytes are a checksum so it would be nice if that was validated too.

**Comment by Technohacker** [comment]
This might take me a while, but I'll try getting it done :)
If anyone gets it done, I'd highly suggest notifying all CryptoNote coins to follow suit as well

**Comment by matt2134711** [comment]
We already have Bouncy Castle as a dependency, which provides all of the cryptographic primitives needed (just keccak-256 IIRC), but I'm not sure if Bitcoin's base58 decoding class can be used as it is slightly different.

I can have a go at implementing this today, if you don't mind.

**Comment by Technohacker** [comment]
Sure, I'll see if I can get an implementation done as well :smile:

**Comment by Technohacker** [comment]
I've made changes to avoid the invalid characters (and made it a utility class for other CryptoNotes). What remains is the checksum. CryptoNote doesn't do decoding like Bitcoin (in blocks of bytes rather than as a whole), and I'm not sure if adding the decoder only for CryptoNote addresses is warrantable. Of course on the other hand there's a risk of putting a wrong (but valid) address.

**Comment by matt2134711** [comment]
The decoder shouldn't be too complicated, the JavaScript on [luigi1111's site](https://xmr.llcoins.net/js/base58.js) is a good reference.

**Comment by ManfredKarrer** [comment]
@matt2134711 @Technohacker We are in the process to remove Bouncy Castle as security provider as with Java 10 they removed support for the extension mechanism (which was used to add the BC signed jar) and we have not found a way yet to include it in the fat jar for the binary otherwise. So we will use the default provider which have less features but covers the features we use atm. I doubt that keccak-256 will be supported there.

**Comment by matt2134711** [comment]
@ManfredKarrer Would it be okay to depend on something like [keccakj](https://github.com/aelstad/keccakj) instead?

**Comment by ManfredKarrer** [comment]
@matt2134711 Is there a jar on a public repo? If so, yes we add that dependency.

**Comment by Technohacker** [comment]
I've (force-)pushed my changes for the new `CryptonoteAddressValidator`. Checksum verification is still pending

**Review by ManfredKarrer — APPROVED** [review]
ACK for first part of validation. 
Validation including the check sum will be delegated to a new PR.

File: assets/src/main/java/bisq/asset/CryptonoteAddressValidator.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with Bisq. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bisq.asset;
+
+/**
+ * {@link AddressValidator} for Base58-encoded Cryptonote addresses.
+ *
+ * @author Chris Beams
```
Comment (Technohacker) [inline assets/src/main/java/bisq/asset/CryptonoteAddressValidator.java:23]: Damn it, I forgot to change this. Could I post a follow-up PR to fix this?

Comment (ManfredKarrer) [reply assets/src/main/java/bisq/asset/CryptonoteAddressValidator.java:23]: What did u forget? Yes feel free for another PR.

Comment (Technohacker) [reply assets/src/main/java/bisq/asset/CryptonoteAddressValidator.java:23]: The author :sweat_smile:

**Pull request merged**